src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -43,10 +43,10 @@ public CSharpConvertLinqQueryToForEachProvider()
         /// <summary>
         /// Finds a node for the span and checks that it is either a QueryExpressionSyntax or a QueryExpressionSyntax argument within ArgumentSyntax.
         /// </summary>
-        protected override QueryExpressionSyntax FindNodeToRefactor(SyntaxNode root, TextSpan span)
+        protected override async Task<QueryExpressionSyntax> FindNodeToRefactorAsync(Document document, TextSpan selection, CancellationToken cancellationToken)
         {
-            var node = root.FindNode(span);
-            return node as QueryExpressionSyntax ?? (node is ArgumentSyntax argument ? argument.Expression as QueryExpressionSyntax : default);
+            var refactoringHelperService = document.GetLanguageService<IRefactoringHelpersService>();
+            return await refactoringHelperService.TryGetSelectedNodeAsync<QueryExpressionSyntax>(document, selection, refactoringHelperService.ExtractNodeFromDeclarationAndAssignment<QueryExpressionSyntax>, cancellationToken).ConfigureAwait(false);
```

Comment:
can you wrap this honking big line?

Reply:
i agree with manish.  i think it should be an implementation detail.  i don't think teh caller needs to call the secondary helper.

Reply:
I would wait with making the extractor default until I move more refactorings to this system and we see it's really the default one.

Reply:
> I don't think this should be part of the interface. It should be an inner impl detail. Basically, inside the service we'll say "it's common to select this larger node while still intending to select this sub part. so we check for that here blah blahblah"

Reply:
Ok, I'm convinced, will update tomorrow.